proposals/gc/MVP.md

```diff
@@ -565,3 +750,215 @@ See [GC JS API document](MVP-JS.md) .
 * Provide functionality to generate fresh, non-canonical RTTs?
 
 * Provide a way to make data types non-eq, especially immutable ones?
+
+
+
+## Appendix: Formal Rules
+
+### Validity
+
+#### Type Indices (`C |- <typeidx> ok`)
+
+```
+C(x) = ct
+---------
+C |- x ok
+```
+
+#### Value Types (`C |- <valtype> ok`)
+
+```
+
+-----------
+C |- i32 ok
+
+C |- x ok
+-------------
+C |- ref x ok
+```
+
+...and so on.
+
+
+#### Structural Types (`C |- <strtype> ok`)
+```
+(C |- t1 ok)*
+(C |- t2 ok)*
+--------------------
+C |- func t1* t2* ok
+
+(C |- ft ok)*
+------------------
+C |- struct ft* ok
+
+C |- ft ok
+----------------
+C |- array ft ok
+```
+
+#### Defined Types (`C |- <subtype>* ok(x)`)
+
+```
+C |- st ok
+(C |- st <: expand(C(x)))*
+(x < x')*
+--------------------------
+C |- sub x* st ok(x')
+
+C |- dt ok(x)
+C |- dt'* ok(x+1)
+-------------------
+C |- dt dt'* ok(x)
+```
+
+#### Type Definitions (`C |- <deftype>* -| C'`)
+
+```
+C |- dt ok(|C|)
+---------------
+C |- dt -| C,dt
+
+x = |C|    N = |dt*|-1
+C' = C,(rec dt*).0,...,(rec dt*).N
+C' |- dt* ok(x)
+-------------------------------------
+C |- rec dt* -| C'
+
+C |- td -| C'
+C' |- td'* ok
+---------------
+C |- td td'* ok
+```
+
+#### Instructions (`C |- <instr> : [t1*] -> [t2*]`)
+
+```
+expand(C(x)) = func t1* t2*
+---------------------------------------
+C |- func.call : [t1* (ref x)] -> [t2*]
+
+expand(C(x)) = struct t1^i t t2*
+------------------------------------
+C |- struct.get i : [(ref x)] -> [t]
+```
+
+...and so on
+
+
+### Type Equivalence
+
+#### Type Indices (`C |- <typeidx> == <typeidx'>`)
+
+```
+C |- tie(x) == tie(x')
+----------------------
+C |- x == x'
+
+
+--------------------
+C |- rec.i == rec.i
+```
+
+#### Value Types (`C |- <valtype> == <valtype'>`)
+
+```
+
+---------------
+C |- i32 == i32
+
+C |- x == x'
+null? = null'?
+---------------------------------
+C |- ref null? x == ref null'? x'
+```
+
+...and so on.
+
+#### Field Types (`C |- <fldtype> == <fldtype'>`)
+
+```
+C |- t == t'
+--------------------
+C |- mut t == mut t'
+```
+
+#### Structural Types (`C |- <strtype> == <strtype'>`)
```

Comment:
What is the difference between a structural type and a defined type?

Reply:
I see, a `deftype` is a type section entry defined as `<subtype> | rec <subtype>*`, so type section entries can either be on their own (like the current function types) or inside a `rec` group. A `subtype` is a single type definition, whether it is on its own or in a `rec` group.